crates/ruff_linter/src/rules/perflint/rules/manual_list_comprehension.rs

```diff
@@ -219,14 +219,73 @@ pub(crate) fn manual_list_comprehension(checker: &mut Checker, for_stmt: &ast::S
         return;
     }
 
-    let binding_stmt = binding
-        .statement(checker.semantic())
-        .and_then(|stmt| stmt.as_assign_stmt());
+    // Avoid if the for-loop target is used outside of the for loop, e.g.,
+    //
+    // ```python
+    // for x in y:
+    //     filtered.append(x)
+    // print(x)
+    // ```
+    //
+    // If this were a comprehension, x would no longer have the correct scope:
+    //
+    // ```python
+    // filtered = [x for x in y]
+    // print(x)
+    // ```
+    let last_target_binding = checker
+        .semantic()
+        .lookup_symbol(for_stmt_target_id)
+        .expect("for loop target must exist");
 
+    let mut bindings = [last_target_binding].into_iter().chain(
+        checker
+            .semantic()
+            .shadowed_bindings(checker.semantic().scope_id, last_target_binding)
+            .filter_map(|shadowed| shadowed.same_scope().then_some(shadowed.shadowed_id())),
+    );
+
+    let target_binding_id = bindings
+        .find(|binding_id| {
+            let binding = checker.semantic().binding(*binding_id);
+            binding
+                .statement(checker.semantic())
+                .and_then(Stmt::as_for_stmt)
+                .is_some_and(|stmt| stmt.range == for_stmt.range)
+        })
+        .expect("for target binding must exist");
+    let target_binding = checker.semantic().binding(target_binding_id);
+    // TODO: should this be a HashMap?
+    let shadowed_references: Vec<_> = checker
+        .semantic()
+        .shadowed_bindings(checker.semantic().scope_id, target_binding_id)
+        .flat_map(|shadowed| {
+            let shadowed_binding = checker.semantic().binding(shadowed.shadowed_id());
+            shadowed_binding.references()
+        })
+        .collect();
+
+    drop(bindings);
+
+    if target_binding
+        .references()
+        .filter(|r_id| !shadowed_references.contains(r_id))
+        .map(|reference| checker.semantic().reference(reference))
+        .any(|r| !for_stmt.range.contains_range(r.range()))
+    {
+        return;
+    }
+
```

Comment:
I don't understand what this code is doing. Would you mind adding a comment explaining why looking at the `shadowed_references` is necessary?